bluetooth: host: fix redundant reference count

xiaoliang314 opened the pull request:
bt_hci_cmd_create creates a buffer with a reference count of 1, so there is no need to increase the reference count again. Test found that it caused the buffer not be released.

xiaoliang314 committed:
Redundant reference count

bt_hci_cmd_create creates a buffer with a reference count of 1, so there is no need to increase the reference count again. Test found that it caused the buffer not be released.

carlescufi commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -261,7 +261,7 @@ int bt_hci_cmd_send_sync(u16_t opcode, struct net_buf *buf,
 
 	/* Make sure the buffer stays around until the command completes */
-	net_buf_ref(buf);
+	/* net_buf_ref(buf); */
```
what about the case where the buffer is provided by the caller to `bt_hci_cmd_send_sync()`? you don't know what refcount is in that buf in that case.

xiaoliang314 replied:
When bt_hci_cmd_send_sync() returns an error, the buffer will not be released, I think this is incorrect.

xiaoliang314 replied:
Yeah, maybe you should use other methods to fix this problem.

xiaoliang314 commented:
I think it can be fix like this:

	if (!buf) {
		buf = bt_hci_cmd_create(opcode, 0);
		if (!buf) {
			return -ENOBUFS;
		}
	} else {
		/* Make sure the buffer stays around until the command completes */
		net_buf_ref(buf);
	}

carlescufi commented:
@xiaoliang314 can you update the PR with the code you posted in the review comment? Remember to amend the commit instead of adding a new commit.

carlescufi commented:
@xiaoliang314 you are also missing a signed-off line in your commit message.
Use `git commit --amend -s` to add that.

xiaoliang314 committed:
bluetooth: host: fix redundant reference count

bt_hci_cmd_create creates a buffer with a reference count of 1, so there is no need to increase the reference count again. Test found that it caused the buffer not be released.

xiaoliang314 committed:
bluetooth: host: fix redundant reference count

bt_hci_cmd_create creates a buffer with a reference count of 1, so there is no need to increase the reference count again. Test found that it caused the buffer not be released.

Signed-off-by: Xiao Liang <xiao_l@profact-tech.com>

johnflux commented:
Xiao, why did you close this pull request?

carlescufi commented:
@jhedberg do you  think this change in the code is needed?
